pkgmgmtutils/brew/formula_updater.go

```diff
@@ -0,0 +1,179 @@
+	remoteChangePusher formula_updater_types.ChangePusher,
+) *FormulaUpdater {
+	return &FormulaUpdater{
+		gitClient:              gitClient,
+		remoteShaGetter:        remoteShaGetter,
+		random:                 random,
+		localCloneChangePusher: localCloneChangePusher,
+		remoteChangePusher:     remoteChangePusher,
+	}
+}
+
+type FormulaUpdater struct {
+	gitClient              formula_updater_types.GitClient
+	remoteShaGetter        formula_updater_types.RemoteShaGetter
+	random                 formula_updater_types.Random
+	localCloneChangePusher formula_updater_types.ChangePusher
+	remoteChangePusher     formula_updater_types.ChangePusher
+}
+
+func NewFormulaUpdaterWithDefaults(ctx context.Context) (*FormulaUpdater, error) {
+	client, err := githubutils.GetClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewFormulaUpdater(
+		internal.NewGitClient(client),
+		internal.NewRemoteShaGetter(),
+		internal.NewRandom(),
+		internal.NewLocalCloneChangePusher(),
+		internal.NewRemoteChangePusher(client),
+	), nil
+}
+
+// for each option in the options slice, update the formula using those options
+// the `version` arg here can often be derived from `versionutils.GetReleaseVersionOrExitGracefully()`
+func (f *FormulaUpdater) Update(
+	ctx context.Context,
+	version *versionutils.Version,
+	projectRepoOwner string,
+	projectRepoName string,
+	formulaOptionsList []*formula_updater_types.FormulaOptions,
+) ([]formula_updater_types.FormulaStatus, error) {
+	if len(formulaOptionsList) == 0 {
+		return nil, nil
+	}
+
+	versionStr := version.String()[1:] // skip the leading "v" in the version string
+
+	// Get version tag SHA
+	// GitHub API docs: https://developer.github.com/v3/git/refs/#get-a-reference
+	gitRefSha, err := f.gitClient.GetRefSha(ctx, projectRepoOwner, projectRepoName, "refs/tags/"+versionStr)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get list of release assets from GitHub
+	// GitHub API docs: https://developer.github.com/v3/repos/releases/#get-a-release-by-tag-name
+	releaseAssets, err := f.gitClient.GetReleaseAssetsByTag(ctx, projectRepoOwner, projectRepoName, versionStr)
+	if err != nil {
+		return nil, err
+	}
+
+	perPlatformShas, err := f.getPerPlatformShas(releaseAssets)
+	if err != nil {
+		return nil, err
+	}
+
+	var formulaStatuses []formula_updater_types.FormulaStatus
+	for _, formulaOptions := range formulaOptionsList {
+		status := formula_updater_types.FormulaStatus{}
+
+		// a version is not stable if it has a label, like "rc", "beta", etc.
+		// in either case, silently mark it as updated and continue
+		if formulaOptions.DryRun || (version.Label != "" && !formulaOptions.PublishUnstableVersion) {
+			status.Updated = true
+			formulaStatuses = append(formulaStatuses, status)
+			continue
+		}
+
+		// Suffix branch name with random number to prevent collisions in rebuilding releases
+		branchName := fmt.Sprintf("%s-%s-%d", formulaOptions.FormulaName, version, f.random.Intn(1000))
```

Comment:
not a big fan of the collision prevention logic here, can we use the git commit hash instead?

Reply:
or git commit sha and not push if the branch already exists?